src/compose/service.ts

```diff
@@ -702,24 +726,76 @@ export class Service {
 				log.debug('  Network changes detected');
 			}
 		}
-		return sameNetworks && sameConfig;
+
+		// Check the network mode separetely, as if it is a
+		// service: network mode, the container id needs to be
+		// checked against the running containers
+
+		// When this function is called, it's with the current
+		// state as a parameter and the target as the instance.
+		// We shouldn't rely on that because it's not enforced
+		// anywhere. For that reason we need to consider both
+		// network_modes in the correct way
+		let sameNetworkMode = false;
+		for (const [a, b] of [
+			[this.config.networkMode, service.config.networkMode],
+			[service.config.networkMode, this.config.networkMode],
+		]) {
+			const aMatch = a.match(SERVICE_NETWORK_MODE_REGEX);
+			const bMatch = b.match(SERVICE_NETWORK_MODE_REGEX);
+			if (aMatch != null) {
+				if (bMatch != null) {
+					// They're both service: <name> network modes
+					if (aMatch[1] !== bMatch[1]) {
+						continue;
+					}
```

Comment:
I think this is missing an `else { sameNetworkMode = true; }`?

(I don't think we'll ever have a case where both have `service:` network modes, but for consistency this should be true if they match, shouldn't it?